cmd/sops/common/common.go

```diff
@@ -136,3 +140,203 @@ func DefaultStoreForPath(path string) Store {
+		return nil, err
+	}
+
+	encCtxBug, err := DetectKMSEncryptionContextBug(tree)
+	if err != nil {
+		return nil, err
+	}
+	if encCtxBug {
+		tree, err = FixAWSKMSEncryptionContextBug(opts, tree)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return tree, nil
+}
+
+// FixAWSKMSEncryptionContextBug is used to fix the issue described in https://github.com/mozilla/sops/pull/435
+func FixAWSKMSEncryptionContextBug(opts GenericDecryptOpts, tree *sops.Tree) (*sops.Tree, error) {
+	message := "Up until version 3.3.0 of sops there was a bug surrounding the " +
+		"use of encryption context with AWS KMS." +
+		"\nYou can read the full description of the issue here:" +
+		"\nhttps://github.com/mozilla/sops/pull/435" +
+		"\n\nIf a TTY is detected, sops will ask you if you'd like for this issue to be " +
+		"automatically fixed, which will require re-encrypting the data keys used by " +
+		"each key." +
+		"\n\nIf you are not using a TTY, sops will fix the issue for this run.\n\n"
+	fmt.Println(wordwrap.WrapString(message, 75))
+
+	persistFix := false
+
+	if terminal.IsTerminal(int(os.Stdout.Fd())) {
+		var response string
+		for response != "y" && response != "n" {
+			fmt.Println("Would you like sops to automatically fix this issue? (y/n): ")
+			_, err := fmt.Scanln(&response)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if response == "n" {
+			return nil, fmt.Errorf("Exiting. User responded no.")
+		} else {
+			persistFix = true
+		}
+	}
+
+	dataKey := []byte{}
+	// If there is another key, then we should be able to just decrypt
+	// without having to try different variations of the encryption context.
+	dataKey, err := DecryptTree(DecryptTreeOpts{
+		Cipher:      opts.Cipher,
+		IgnoreMac:   opts.IgnoreMAC,
+		Tree:        tree,
+		KeyServices: opts.KeyServices,
+	})
+	if err != nil {
+		dataKey = RecoverDataKeyFromBuggyKMS(opts, tree)
+	}
+
+	if dataKey == nil {
+		return nil, NewExitError(fmt.Sprintf("Failed to decrypt, meaning there is likely another problem from the encryption context bug: %s", err), codes.ErrorDecryptingTree)
+	}
+
+	errs := tree.Metadata.UpdateMasterKeysWithKeyServices(dataKey, opts.KeyServices)
+	if len(errs) > 0 {
+		err = fmt.Errorf("Could not re-encrypt data key: %s", errs)
+		return nil, err
+	}
+
+	err = EncryptTree(EncryptTreeOpts{
+		DataKey: dataKey,
+		Tree:    tree,
+		Cipher:  opts.Cipher,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// If we are not going to persist the fix, just return the re-encrypted tree.
+	if !persistFix {
+		return tree, nil
+	}
+
+	encryptedFile, err := opts.InputStore.EmitEncryptedFile(*tree)
+	if err != nil {
+		return nil, NewExitError(fmt.Sprintf("Could not marshal tree: %s", err), codes.ErrorDumpingTree)
+	}
+
+	file, err := os.Create(opts.InputPath)
+	if err != nil {
+		return nil, NewExitError(fmt.Sprintf("Could not open file for writing: %s", err), codes.CouldNotWriteOutputFile)
+	}
+	_, err = file.Write(encryptedFile)
+	if err != nil {
+		file.Close()
```

Comment:
uber-nit: could `defer file.Close()` after `os.Create(opts.InputPath)` instead of calling it on each branch

Reply:
I *think* we want to keep it like this, only because directly after this (within this function) we call `LoadEncryptedFile()`.

Reply:
But, I can remove the duplicate calls and call it directly after the `file.Write()`

Reply:
jk, using `defer` works as desired here.